File: src-tauri/src/preferences.rs
//! What a person chose: the interface's palette, its identity colour, its language, and which
//! model the agent is asked for.
//!
//! Four small facts, kept in one file so that they survive a restart, and kept **here** rather
//! than in the webview's own storage, which is somewhere nobody can find, back up or delete
//! with the conventions of their own operating system. They are what the user chose, so they
//! are configuration, so they live in the configuration directory, and the path comes from the
//! resolver rather than from a literal.
//!
//! **The vocabularies are not here.** This side knows that there are four choices and what
//! they are called; it does not know that a palette is `light` or `dark`, that a language is
//! `en` or `es`, or what any model is called. Those lists live where they are already written
//! down — `src/styles/tokens.css` for the two the interface is drawn from, `src/i18n/` for the
//! language, `src/lib/models.ts` for the model — and adding a fourth accent, a third language
//! or another model must not mean editing Rust as well. Anything this file does not recognise
//! is a string it stores and hands back, and the interface is what narrows it.
//!
//! Nothing in here is personal data and nothing ever will be: a palette, a colour, a language
//! and the name of a model are choices about how a program behaves, not facts about a person.

use std::{fs, path::PathBuf};

use log::{info, warn};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

/// The file the choices are kept in, inside the configuration directory.
const FILE: &str = "preferences.json";

/// What a person chose, with `None` wherever they have chosen nothing.
///
/// Absent is not the same as a default written down: it is the interface that decides what
/// "nobody has chosen a language" means, and its answer — the language the device asks for —
/// is not a value this side could have stored in advance.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(default)]
pub struct Preferences {
    /// The palette, or `None` to follow the operating system.
    pub theme: Option<String>,
    /// The identity colour, or `None` for the one the application icon wears.
    pub accent: Option<String>,
    /// The language, or `None` for the one the device asks for.
    pub language: Option<String>,
    /// The model the agent is asked for, or `None` for the agent's own default.
    ///
    /// Absent means the variable is not set at all when the agent starts, so the agent decides.
    /// The alternative — a default written down here — would put the vocabulary back on this
    /// side through the door the paragraph above closes.
    pub model: Option<String>,
}

/// Where the file is, or nothing when the platform will not say.
///
/// A failure here is not worth ending anything over: an application that cannot find its
/// configuration directory runs with the defaults, which is what it does on a first launch.
fn file(app: &AppHandle) -> Option<PathBuf> {
    match app.path().app_config_dir() {
        Ok(directory) => Some(directory.join(FILE)),
        Err(error) => {
            warn!("preferences_not_located reason={error}");
            None
        }
    }
}

/// Reads what was stored, or the defaults when nothing was.
///
/// Every failure lands on the same answer, and deliberately: a missing file is a first launch,
/// and a file this build cannot parse is one written by a version that meant something else by
/// it. Neither is a reason to refuse to draw a screen, and both leave the person with an
/// interface they can set again.
fn read(app: &AppHandle) -> Preferences {
    let Some(path) = file(app) else {
        return Preferences::default();
    };

    let Ok(text) = fs::read_to_string(&path) else {
        return Preferences::default();
    };

    serde_json::from_str(&text).unwrap_or_else(|error| {
        warn!("preferences_not_understood reason={error}");
        Preferences::default()
    })
}

/// Writes the choices out, and says in the log when it could not.
///
/// # Errors
///
/// Returns whatever the filesystem raised: the directory not being creatable, or the file not
/// being writable.
fn write(app: &AppHandle, preferences: &Preferences) -> std::io::Result<()> {
    let Some(path) = file(app) else {
        return Ok(());
    };

    if let Some(directory) = path.parent() {
        fs::create_dir_all(directory)?;
    }

    let text = serde_json::to_string_pretty(preferences)
        .map_err(|error| std::io::Error::other(error.to_string()))?;

    fs::write(path, text)
}

/// What a person has chosen so far.
#[tauri::command]
pub fn preferences(app: AppHandle) -> Preferences {
    read(&app)
}

/// Stores the choices, and returns what is stored afterwards.
///
/// The answer is read back rather than echoed, for the reason `open_at_login` reads its own
/// setting back: a caller can then tell a change from a refusal by comparing the two, and a
/// control that slid across while nothing was written is the one failure worth avoiding.
///
/// The whole set is written at once rather than one field at a time, so that two choices made
/// in quick succession cannot race each other into a half-written file.
#[tauri::command]
pub fn set_preferences(app: AppHandle, preferences: Preferences) -> Preferences {
    if let Err(error) = write(&app, &preferences) {
        warn!("preferences_not_stored reason={error}");
    } else {
        info!("preferences_stored");
    }

    read(&app)
}

File: src/lib/preferences.ts
/**
 * What a person chose, and what it takes to keep it.
 *
 * Four choices — the palette, the identity colour, the language, and which model the agent is
 * asked for — held in one file by the Rust side, which is where a file this application writes
 * belongs (`.agents/rules/storage-and-logs.md`). Nothing here knows what any of them
 * *mean*: this file is the store, and the vocabularies live where they are already written
 * down — `@/lib/appearance` for the two the interface is drawn from, `@/i18n` for the
 * language, `@/lib/models` for the model.
 *
 * The choices are read **once**, before anything is drawn, and held here afterwards. That is
 * what lets a screen ask what the palette is without waiting: a settings screen whose controls
 * arrive a moment after the rest of it is a settings screen that flickers into its own state.
 */

import { invoke } from "@tauri-apps/api/core";

/**
 * The four choices, as they are stored.
 *
 * Every one of them may be `null`, and `null` is not a default written down: it is nobody
 * having chosen. What the absence means is the reader's — the operating system's palette, the
 * icon's own colour, the device's language — and none of those is a value that could have been
 * stored in advance.
 */
export interface Preferences {
  /** The palette a person asked for, or `null` while they have asked for none. */
  theme: string | null;
  /** The identity colour a person asked for, or `null` while they have asked for none. */
  accent: string | null;
  /** The language a person asked for, or `null` while they have asked for none. */
  language: string | null;
  /**
   * The model the agent is asked for, or `null` while nobody has chosen one.
   *
   * `null` is not a default written down here either: it means the agent is told nothing and
   * uses its own, which is a value this side deliberately does not know — see
   * `@/lib/models`, which holds the list this side *can* name.
   */
  model: string | null;
}

/** Nobody has chosen anything. What a first launch has, and what a failed read falls back to. */
const NOTHING_CHOSEN: Preferences = {
  theme: null,
  accent: null,
  language: null,
  model: null,
};

/** What was read the last time anybody asked the Rust side. */
let held: Preferences = NOTHING_CHOSEN;

/**
 * Reads the choices from disk and holds them.
 *
 * Called once, before the first render. A failure is the defaults rather than an error: the
 * question could not be asked, and an interface drawn in the palette nobody chose is a working
 * interface, where one that refused to draw at all is not.
 *
 * @returns What is now held, which is what every later call to {@link preferences} answers.
 */
export async function loadPreferences(): Promise<Preferences> {
  try {
    held = await invoke<Preferences>("preferences");
  } catch {
    held = NOTHING_CHOSEN;
  }

  return held;
}

/** What a person has chosen, as of the last read. */
export function preferences(): Preferences {
  return held;
}

/**
 * Stores a choice and returns what is stored afterwards.
 *
 * The answer comes back from the store rather than echoing the request, which is what lets a
 * caller tell a change from a refusal by comparing the two — the same reason
 * `@/lib/openAtLogin` reads its setting back. A choice that could not be written is a control
 * that goes back to where it was, rather than one that claims something nobody kept.
 *
 * @param chosen - The choices that changed. The rest are carried over unread.
 * @returns Everything that is stored now.
 */
export async function choose(chosen: Partial<Preferences>): Promise<Preferences> {
  const wanted = { ...held, ...chosen };

  try {
    held = await invoke<Preferences>("set_preferences", { preferences: wanted });
  } catch {
    // The command could not be reached at all. Read them back the ordinary way rather than
    // reporting a choice nobody kept.
    return await loadPreferences();
  }

  return held;
}
